### PR TITLE
Creating a view-recipe template

### LINKED ISSUE #89: Create a template for the View-Recipe page
We need a template for the page where users will be able to view a recipe,
so we could work on its associated view(s) and also provide a consistent standard
for previews in the Create/Edit-Recipe page.

### PR DESCRIPTION
**Pull request by GilShechter**
### Description

This PR is creating a template for presenting a single recipe.
The template is built to show all the recipe's relevant data i.e image, description, rating, cooking time, categories, author, date posted, and ingredients.

Close #89

### What's new

- CSS file and HTML file for the template
- View function
- URL route
- Test

### What's **not** included yet

- The comments feature
- The add/edit rating feature
- Functionality for the category buttons
- Option to edit the recipe if you're the author

### What it looks like

view_recipe

### CODE REVIEW

File: tasties/urls.py [hunk 1]
```diff
@@ -24,5 +24,7 @@
     path('', views.recipes, name="recipes"),
     path('login/', views.login_user, name="login"),
     path('register/', views.register, name="register"),
-    path('logout/', views.logout_user, name="logout")
+    path('logout/', views.logout_user, name="logout"),
+    # path('single_recipe/<int:recipe_id>/', views.single_recipe, name="single_recipe"),
+    path('single_recipe/', views.single_recipe, name="single_recipe")
```
Comment (efraimvis) [inline tasties/urls.py:29]: I think `single_recipe` might not be the most descriptive name for the page. Maybe consider `view_recipe`?

File: tasties_app/templates/tasties_app/single_recipe.html [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            <h2>{{ recipe.description }}</h2>
+            <div class="recipe-header-info">
+                <span class="author-name">by @{{ recipe.author_id }}</span>
+                <span class="recipe-date">{{ recipe.publication_date }}</span>
+            </div>
+        </div>
+
+        <div class="categories">
+            {% for category in categories %}
+                <button class="btn btn-light" name="{{ category.category_name }}">{{ category.category_name }}</button>
```
Comment (efraimvis) [inline tasties_app/templates/tasties_app/single_recipe.html:20]: I recommend a slightly darker/different color for the category tags, so they stand out more against the background.

File: tasties_app/views.py [hunk 3]
```diff
@@ -61,3 +61,14 @@ def register(request):
 
     context = {'form': form}
     return render(request, 'tasties_app/register.html', context)
+
+
+@login_required(login_url='login')
+def single_recipe(request, recipe=None):
+    if not recipe:
+        recipe = Recipe.objects.first()
+    ingredients = Ingredient.objects.filter(recipe_id=recipe)
```
Comment (efraimvis) [inline tasties_app/views.py:70]: You might not even have to do this. I think you can have: `ingredients = recipe.ingredient_set.all()` and get a slightly more readable assignment, maybe possibly but not sure even a slightly more efficient DB query.

File: tasties_app/views.py [hunk 4]
```diff
@@ -61,3 +61,14 @@ def register(request):
...
+    rating = Rating.objects.filter(recipe_id=recipe).aggregate(Avg('rating'))['rating__avg']
```
Comment (efraimvis) [inline tasties_app/views.py:71]: Here you can definitely utilize `rating = recipe.rating_set.aggregate(Avg('rating__rating'))`

File: tasties_app/templates/tasties_app/single_recipe.html [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                        <span class="recipe-rating-icon">{% bs_icon 'star-fill' %}</span>
+                        <span class="recipe-rating-text">{{ rating }}</span>
+                    </div>
+                </div>
+            <hr>
+                <ul class="ingredients-list list-group list-group">
+                    <h3>Ingredients</h3>
+                    {% for ingredient in ingredients %}
+                        <li class="ingredient">
+                            {{ ingredient }}
```
Comment (efraimvis) [inline tasties_app/templates/tasties_app/single_recipe.html:43]: Here you're just calling the `Ingredient` model's `__str__` method; This is fine for the time being, but please notice
that we will need to modify/augment this in the future. For example, I think that we should (in the future) omit the word "Whole" from the string, should it be the measurement unit - think "5 Chicken breasts Whole" etc. We would also like to do unit conversion, and that will have to happen somewhere.

File: tasties_app/views.py (hunk 3 above)
Comment (efraimvis) [inline tasties_app/views.py:70]: Another thing that will need to be in here in the future is automatic/selective unit conversion. For example, when we exceed some "reasonable" value, let's say we have "1052 grams" of some ingredient, we may convert to "1.052 Kilograms", we might as well do it in the view, since we may also want to implement some "Metric/Imperial" switch - should we decide to pursue that.

File: tasties_app/views.py [hunk 6]
```diff
@@ -61,3 +61,14 @@ def register(request):
 
     context = {'form': form}
     return render(request, 'tasties_app/register.html', context)
+
+
+@login_required(login_url='login')
+def single_recipe(request, recipe=None):
+    if not recipe:
+        recipe = Recipe.objects.first()
```
Comment (efraimvis) [inline tasties_app/views.py:69]: This is temporary, I know, but risky... Maybe worth your effort to put a comment over this line saying it should be amended in the future.

**Review by efraimvis — CHANGES_REQUESTED** [review]
Great job overall - excellent design. Please see my notes, the biggest one I think is that this whole concept should be given a more meaningful name than `single_recipe`, which I think is not a great description, and should probably become something like `view_recipe`.

File: tasties_app/templates/tasties_app/single_recipe.html [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+{% load static %}
+{% block css %}
+    {% load bootstrap_icons %}
+    <link rel="stylesheet" type="text/css" href="{% static 'css/single_recipe.css' %}" />
+{% endblock %}
+{% block content %}
+    <div class="recipe-body">
+        <div class="recipe-header">
+            <h1>{{ recipe.title }}</h1>
+            <h2>{{ recipe.description }}</h2>
```
Comment (efraimvis) [inline tasties_app/templates/tasties_app/single_recipe.html:11]: You know what, I would also move the description lower down (between the recipe picture and the directions), and also not make it an h2, as that might be too bold (but it might be the right choice to leave it as an h2 because of accessibility concerns, just modify the font-weight for h2 in the CSS). See [this](https://www.epicurious.com/recipes/food/views/air-fryer-brussels-sprouts-salad-grain-bowl#:~:text=It%20might%20be,if%20you%20like.) as an example that descriptions can be a bit verbose.

File: tasties_app/templates/tasties_app/single_recipe.html [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+            <div class="recipe-header-info">
+                <span class="author-name">by @{{ recipe.author_id }}</span>
```
Comment (efraimvis) [inline tasties_app/templates/tasties_app/single_recipe.html:13]: Maybe the word "by" should start with a capital letter, "By".

File: tasties_app/templates/tasties_app/single_recipe.html [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+                <span class="recipe-date">{{ recipe.publication_date }}</span>
```
Comment (efraimvis) [inline tasties_app/templates/tasties_app/single_recipe.html:14]: I would add "Published on: " before the publication date.

File: tasties_app/views.py (hunk 3 above)
Comment (GilShechter) [reply tasties_app/views.py:70]: As for this comment and the one above - At this moment I doubt that we will have time for that.
Let's continue with the tasks we agreed on for now.
If we'll finish earlier than expected we could try adding that conversion functionality.

File: tasties_app/views.py (hunk 6 above)
Comment (GilShechter) [reply tasties_app/views.py:69]: I'll comment but it's gonna change very soon. Working on it.

File: tasties_app/templates/tasties_app/single_recipe.html (hunk 9 above)
Comment (GilShechter) [reply tasties_app/templates/tasties_app/single_recipe.html:14]: Liked it better without it, the same as in the example you sent me.

**Review by ohadrevach — APPROVED** [review]
Great work!! , looking very good

**Review by efraimvis — DISMISSED** [review]
Good for now!

File: tasties/urls.py [hunk 10]
```diff
@@ -24,5 +24,7 @@
     path('', views.recipes, name="recipes"),
     path('login/', views.login_user, name="login"),
     path('register/', views.register, name="register"),
-    path('logout/', views.logout_user, name="logout")
+    path('logout/', views.logout_user, name="logout"),
+    # path('view_recipe/<int:recipe_id>/', views.single_recipe, name="single_recipe"),
```
Comment (natifridman) [inline tasties/urls.py:28]: So this comment going replaced in another PR?

File: tasties_app/templates/tasties_app/view_recipe.html [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+                <span class="author-name">By @{{ recipe.author_id }}</span>
+                <span class="recipe-date">{{ recipe.publication_date }}</span>
+            </div>
+        </div>
+
+        <div class="categories">
+            {% for category in categories %}
+                <button class="btn btn-outline" name="{{ category.category_name }}">{{ category.category_name }}</button>
+            {% endfor %}
+
```
Comment (natifridman) [inline tasties_app/templates/tasties_app/view_recipe.html:22]: I think this newline is redundant

**Review by natifridman — CHANGES_REQUESTED** [review]
Maybe we can add more tests here, what about a test for an unauthorized user?

File: tasties/urls.py (hunk 10 above)
Comment (GilShechter) [reply tasties/urls.py:28]: Yes, when we'll have the ability to enter a recipe from the recipe's menu.
I'll remove it for now.

**Comment by GilShechter** [comment]
Added a test to verify that an unauthorized user who tries to access this URL is being redirected to the login page.

File: tasties_app/views.py (hunk 6 above)
Comment (natifridman) [reply tasties_app/views.py:69]: Why not redirect to somewhere else in case there is no recipe?

Comment (GilShechter) [reply tasties_app/views.py]: Changes the method to redirect to the `recipes` page if no recipe is given.

File: tasties_app/tests/test_recipes_view.py [hunk 12]
```diff
@@ -11,3 +11,14 @@ def test_recipes_view(self, client, signed_up_credentials):
         response = client.post('/login/', data={'username': VALID_USER, 'password': VALID_PASSWORD})
         assert response.status_code == 302
         assert response.url == '/'
+
+    def test_view_recipe(self, client, signed_up_credentials):
+        client.post('/login/', data={'username': VALID_USER, 'password': VALID_PASSWORD})
+        response = client.get('/view_recipe/')
+        assert response.status_code == 302
```
Comment (natifridman) [inline tasties_app/tests/test_recipes_view.py:18]: Just want to validate, another test to check status code 200 is going to be part of another PR?

Comment (GilShechter) [reply tasties_app/tests/test_recipes_view.py:18]: Yes, because at the moment it's not accessible from the app.
But in the next PR, I will add a test that shows a successful  HTTP response.

Comment (natifridman) [reply tasties_app/tests/test_recipes_view.py:18]: Great, thanks!

**Review by natifridman — DISMISSED** [review]
LGTM

**Review by GalOved — DISMISSED** [review]
Looking good for me

**Comment by liornoy** [comment]
@GilShechter I see that the path of the single recipe is: "view_recipe/"
It's more common and conventional to use the pattern of "recipes/<recipe_id>"
for displaying the page of details of an object in REST API.
Is it going to change when @GalOved will work on the Backend?

**Comment by GilShechter** [comment]
> @GilShechter I see that the path of the single recipe is: "view_recipe/" It's more common and conventional to use the pattern of "recipes/<recipe_id>" for displaying the page of details of an object in REST API. Is it going to change when @GalOved will work on the Backend?

Yes, you will see it in the next PR when the request will come from the recipes menu.

**Review by efraimvis — DISMISSED** [review]

**Comment by liornoy** [comment]
Sounds good, please squash the commits, and I'll merge it.

**Review by efraimvis — APPROVED** [review]

**Comment by GilShechter** [comment]
Thanks. Commits squashed.

**Review by natifridman — APPROVED** [review]
